In [1]:
!pwd

/Users/kpadhikari/Desktop/BigFls/CLAS12/GitProj/KPAdhikari/PythonStuff/IPython_Jupyter_Notebooks


Ref: https://packaging.python.org/tutorials/installing-packages/

### Installing Packages

This section covers the basics of how to install Python [packages](https://packaging.python.org/glossary/#term-distribution-package).

It’s important to note that the term “package” in this context is being used as a synonym for a [distribution](https://packaging.python.org/glossary/#term-distribution-package) (i.e. a bundle of software to be installed), not to refer to the kind of [package](https://packaging.python.org/glossary/#term-import-package) that you import in your Python source code (i.e. a container of modules). It is common in the Python community to refer to a [distribution](https://packaging.python.org/glossary/#term-distribution-package) using the term “package”. Using the term “distribution” is often not preferred, because it can easily be confused with a Linux distribution, or another larger software distribution like Python itself.



Contents

    Requirements for Installing Packages

        Ensure you can run Python from the command line

        Ensure you can run pip from the command line

        Ensure pip, setuptools, and wheel are up to date

        Optionally, create a virtual environment

    Creating Virtual Environments

    Use pip for Installing

    Installing from PyPI

    Source Distributions vs Wheels

    Upgrading packages

    Installing to the User Site

    Requirements files

    Installing from VCS

    Installing from other Indexes

    Installing from a local src tree

    Installing from local archives

    Installing from other sources

    Installing Prereleases

    Installing Setuptools “Extras”



### Requirements for Installing Packages

This section describes the steps to follow before installing other Python packages.
### Ensure you can run Python from the command line

Before you go any further, make sure you have Python and that the expected version is available from your command line. You can check this by running:
```sh
    python --version
```
You should get some output like Python 3.6.3. If you do not have Python, please install the latest 3.x version from python.org or refer to the Installing Python section of the Hitchhiker’s Guide to Python.

**Note:** If you’re a newcomer and you get an error like this:
```sh
>>> python --version
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'python' is not defined
```
It’s because this command and other suggested commands in this tutorial are intended to be run in a shell (also called a terminal or console). See the Python for Beginners [getting started tutorial](https://opentechschool.github.io/python-beginners/en/getting_started.html#what-is-python-exactly) for an introduction to using your operating system’s shell and interacting with Python.

**Note:** If you’re using an enhanced shell like IPython or the Jupyter notebook, you can run system commands like those in this tutorial by prefacing them with a ! character:
```sh
    In [1]: import sys
            !{sys.executable} --version
    Python 3.6.3
```
It’s recommended to write {sys.executable} rather than plain python in order to ensure that commands are run in the Python installation matching the currently running notebook (which may not be the same Python installation that the python command refers to).

**Note:** Due to the way most Linux distributions are handling the Python 3 migration, Linux users using the system Python without creating a virtual environment first should replace the python command in this tutorial with python3 and the pip command with `pip3 --user`. Do not run any of the commands in this tutorial with `sudo`: if you get a permissions error, come back to the section on creating virtual environments, set one up, and then continue with the tutorial as written.


In [2]:
!python --version

Python 2.7.16


In [6]:
import sys
#kp: to see various options after 'sys.', type 'sys.' and press tab.
print(sys.executable)
print(sys.argv)
print(sys.abiflags)
print(sys.api_version)
!{sys.executable} --version

/usr/local/opt/python/bin/python3.7
['/usr/local/lib/python3.7/site-packages/ipykernel_launcher.py', '-f', '/Users/kpadhikari/Library/Jupyter/runtime/kernel-73351161-ddd6-40e5-8883-b90e8fc9b848.json']
m
1013
Python 3.7.4


### Ensure you can run pip from the command line

Additionally, you’ll need to make sure you have [pip](https://packaging.python.org/key_projects/#pip) available. You can check this by running:
```sh
pip --version
```
If you installed Python from source, with an installer from [python.org](https://python.org/), or via [Homebrew](https://brew.sh/) you should already have pip. If you’re on Linux and installed using your OS package manager, you may have to install pip separately, see [Installing pip/setuptools/wheel with Linux Package Managers](https://packaging.python.org/guides/installing-using-linux-tools/).

If pip isn’t already installed, then first try to bootstrap it from the standard library:
```sh
python -m ensurepip --default-pip
```
If that still doesn’t allow you to run pip:

* Securely Download [get-pip.py](https://bootstrap.pypa.io/get-pip.py) [[1]](https://packaging.python.org/tutorials/installing-packages/#id7)
* Run python `get-pip.py`. [[2](https://packaging.python.org/tutorials/installing-packages/#id8)] This will install or upgrade pip. Additionally, it will install [setuptools](https://packaging.python.org/key_projects/#setuptools) and [wheel](https://packaging.python.org/key_projects/#wheel) if they’re not installed already.

    **Warning:** Be cautious if you’re using a Python install that’s managed by your operating system or another package manager. get-pip.py does not coordinate with those tools, and may leave your system in an inconsistent state. You can use `python get-pip.py --prefix=/usr/local/` to install in `/usr/local` which is designed for locally-installed software.



In [7]:
!pip --version

/bin/sh: pip: command not found


In [8]:
!pip3 --version

pip 19.1.1 from /usr/local/lib/python3.7/site-packages/pip (python 3.7)


In [9]:
!python -m ensurepip --default-pip

Looking in links: /var/folders/xx/3nmz3j457b99xd1r8kfj0j6r0000gn/T/tmpw8szjj
Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/Library/Python/2.7/site-packages/pip'
Consider using the `--user` option or check the permissions.



### Ensure pip, setuptools, and wheel are up to date

While pip alone is sufficient to install from pre-built binary archives, up to date copies of the `setuptools` and `wheel` projects are useful to ensure you can also install from source archives:
```sh
python -m pip install --upgrade pip setuptools wheel
```
### Optionally, create a virtual environment

See section below for details, but here’s the basic venv 3 command to use on a typical Linux system:
```sh
python3 -m venv tutorial_env
source tutorial_env/bin/activate
```
This will create a new virtual environment in the `tutorial_env `subdirectory, and configure the current shell to use it as the default `python` environment.


In [11]:
!python3 -m pip install --upgrade pip setuptools wheel

Requirement already up-to-date: pip in /usr/local/lib/python3.7/site-packages (20.1.1)
Requirement already up-to-date: setuptools in /usr/local/lib/python3.7/site-packages (49.2.0)
Requirement already up-to-date: wheel in /usr/local/lib/python3.7/site-packages (0.34.2)


### Creating Virtual Environments

Python “Virtual Environments” allow Python [packages](https://packaging.python.org/glossary/#term-distribution-package) to be installed in an isolated location for a particular application, rather than being installed globally. If you are looking to safely install global command line tools, see [Installing stand alone command line tools](https://packaging.python.org/guides/installing-stand-alone-command-line-tools/).

Imagine you have an application that needs version 1 of LibFoo, but another application requires version 2. How can you use both these applications? If you install everything into /usr/lib/python3.6/site-packages (or whatever your platform’s standard location is), it’s easy to end up in a situation where you unintentionally upgrade an application that shouldn’t be upgraded.

Or more generally, what if you want to install an application and leave it be? If an application works, any change in its libraries or the versions of those libraries can break the application.

Also, what if you can’t install packages into the global site-packages directory? For instance, on a shared host.

In all these cases, virtual environments can help you. They have their own installation directories and they don’t share libraries with other virtual environments.

Currently, there are two common tools for creating Python virtual environments:

* [venv](https://docs.python.org/3/library/venv.html) is available by default in Python 3.3 and later, and installs [pip](https://packaging.python.org/key_projects/#pip) and [setuptools](https://packaging.python.org/key_projects/#setuptools) into created virtual environments in Python 3.4 and later.
* [virtualenv](https://packaging.python.org/key_projects/#virtualenv) needs to be installed separately, but supports Python 2.7+ and Python 3.3+, and **pip, setuptools** and [wheel](https://packaging.python.org/key_projects/#wheel) are always installed into created virtual environments by default (regardless of Python version).

The basic usage is like so:

Using venv:
```sh
python3 -m venv <DIR>
source <DIR>/bin/activate
```
Using virtualenv:
```sh
virtualenv <DIR>
source <DIR>/bin/activate
```
For more information, see the [venv](https://docs.python.org/3/library/venv.html) docs or the [virtualenv](http://virtualenv.pypa.io/) docs.

The use of `source` under Unix shells ensures that the virtual environment’s variables are set within the current shell, and not in a subprocess (which then disappears, having no useful effect).

In both of the above cases, Windows users **should _not_ use** the `source` command, but should rather run the activate script directly from the command shell like so:
```sh
<DIR>\Scripts\activate
```
Managing multiple virtual environments directly can become tedious, so the [dependency management tutorial](https://packaging.python.org/tutorials/managing-dependencies/#managing-dependencies) introduces a higher level tool, [Pipenv](https://packaging.python.org/key_projects/#pipenv), that automatically manages a separate virtual environment for each project and application that you work on.
